components/net/websocket_loader.rs

```diff
@@ -3,63 +3,107 @@
+    fn build_request(&mut self, url: &Url) -> WebSocketResult<Request> {
+        debug!("Handler is building request from {}.", url);
+        try!(parse_url(url).map_err(|e| WebSocketError::new(WebSocketErrorKind::Protocol, e.description().to_owned())));
+        let mut req = try!(Request::from_url(url));
+        req.headers_mut().push(("Origin".to_string(), self.origin.as_bytes().to_owned()));
+        req.headers_mut().push(("Host".to_string(), format!("{}", self.host).as_bytes().to_owned()));
+        for protocol in self.protocols {
+            req.add_protocol(protocol);
         };
+        let mut headers = WebSocketHeaders::new(req.headers().clone()).as_hyper_headers();
+        set_request_cookies(self.resource_url, &mut headers, self.cookie_jar);
+        if let Some(cookie) = headers.get_raw("Cookie") {
+            req.headers_mut().push(("Cookie".to_string(), cookie[0].to_owned()));
+        }
+        Ok(req)
     }
 
-    let headers = response.headers.clone();
-    let (sender, receiver) = response.begin().split();
-    Ok((headers, sender, receiver))
+    fn on_open(&mut self, shake: Handshake) -> WebSocketResult<()> {
+        let headers = WebSocketHeaders::new(shake.response.headers().clone());
+        set_cookies_from_headers(self.resource_url, &headers.as_hyper_headers(), self.cookie_jar);
+        let _ = self.event_sender.send(
+            WebSocketNetworkEvent::ConnectionEstablished(headers, self.protocols.to_owned()));
+        Ok(())
+    }
+
+    fn on_message(&mut self, message: Message) -> WebSocketResult<()> {
+        let message = match message {
+            Message::Text(message) => MessageData::Text(message),
+            Message::Binary(message) => MessageData::Binary(message),
+        };
+        let _ = self.event_sender.send(WebSocketNetworkEvent::MessageReceived(message));
+        Ok(())
+    }
+
+    fn on_error(&mut self, err: WebSocketError) {
+        debug!("Error in WebSocket communication: {:?}", err);
+        let _ = self.event_sender.send(WebSocketNetworkEvent::Fail);
+    }
+
+    fn on_response(&mut self, res: &Response) -> WebSocketResult<()> {
+        let protocol_in_use = try!(res.protocol());
+        if let Some(protocol_name) = protocol_in_use {
+            let protocol_name = protocol_name.to_lowercase();
+            if self.protocols.is_empty() && !self.protocols.iter().any(|p| protocol_name == (*p).to_lowercase()) {
+                let error = WebSocketError::new(WebSocketErrorKind::Protocol,
+                                                "Protocol in Use not in client-supplied protocol list");
+                return Err(error);
+            }
+        }
+        Ok(())
+    }
 
+    fn on_close(&mut self, code: CloseCode, reason: &str) {
+        debug!("Connection closing due to ({:?}) {}", code, reason);
+        let reason = if code == CloseCode::Status {
+            "".to_owned()
+        } else {
+            reason.to_string()
+        };
```

Comment:
We should be echoing the message back to the server here in [some cases](https://tools.ietf.org/html/rfc6455#section-1.4); https://docs.rs/ws/0.6.0/ws/trait.Handler.html#method.on_close doesn't show this being handled automatically.